### PR TITLE
feat(elm-challenge): Solution for the first part of the challenge

### PR DESCRIPTION
**Pull request by vpjapp**
Hello,

Here's my attempt at cracking the tree browser nut. Hope you like it :)

### CODE REVIEW

**Comment by vpjapp** [comment]
READY

File: elm-tooling.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+    "entrypoints": [
+        "./src/Main.elm"
+    ],
+    "tools": {
+        "elm": "0.19.1",
+        "elm-format": "0.8.5",
+        "elm-json": "0.2.10",
+        "elm-test-rs": "1.0.0"
+    }
+}
```
Comment (Janiczek) [inline elm-tooling.json:11]: Nice to see `elm-tooling` used 🙂

File: package.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "elm-tooling": "^1.3.0",
+    "node-typescript": "^0.1.3",
+    "parcel": "^2.0.0-beta.2"
+  },
+  "devDependencies": {
+    "elm": "^0.19.1-5"
+  },
+  "scripts": {
+    "start": "parcel src/index.html",
+    "test": "elm-test-rs"
```
Comment (Janiczek) [inline package.json:20]: Interesting that you don't need
```js
    "postinstall": "elm-tooling install"
```
here. We usually do 🤷 😄

File: tests/AudienceTest.elm [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                    |> (\res ->
+                            case res of
+                                Result.Ok list ->
+                                    list
+
+                                Result.Err _ ->
+                                    []
+                       )
+                    |> List.length
+                    |> Expect.equal 212
```
Comment (Janiczek) [inline tests/AudienceTest.elm:24]: Nice idea to make sure the JSON gets parsed correctly.

File: src/Main.elm [hunk 4]
```diff
@@ -1,10 +1,285 @@
...
+
+
+class : String -> Html.Attribute msg
+class =
+    Html.Attributes.class
+
+
+style : String -> String -> Html.Attribute msg
+style =
+    Html.Attributes.style
```
Comment (Janiczek) [inline src/Main.elm:254]: These probably could have been just exposed unqualified in the imports?
```elm
import Html.Attributes exposing (class, style)
```

File: src/Main.elm [hunk 5]
```diff
@@ -1,10 +1,285 @@
...
+    let
+        classes =
+            "node"
+                ++ (if current then
+                        " current"
+
+                    else
+                        ""
+                   )
+    in
```
Comment (Janiczek) [inline src/Main.elm:273]: There is this nice function `Html.Attributes.classList`:
```elm
classList
  [ ("node", True)
  , ("current", current)
  ]
```
that could replace this manual conditional string appending

**Review by Janiczek — COMMENTED** [review]
Overall pretty nice!

Nice that you used Tree.Zipper and made it do the work as needed (JIT), not upfront (AOT).

As you mentioned, it might not make a difference for this particular dataset, but we have had ~6MB worth of data for some other entity type that needed the same tree approach and we also switched to lazy to accomodate for that.

I left some suggestions in the code :)

**Comment by Janiczek** [comment]
Oh and it's probably a bad form to commit the NPM / Yarn lockfiles (`package-lock.json` in your case). But it doesn't matter much here :)

**Comment by vpjapp** [comment]
Hi,

I meant to comment on these earlier, but kinda forgot 😅.

> Oh and it's probably a bad form to commit the NPM / Yarn lockfiles

How come? Isn't it the whole point of the lock file that each team member
can have the exact same versions of each dependency? You know js-ecosystem,
even a patch update may break everything 🙈

> Interesting that you don't need
> "postinstall": "elm-tooling install"

Good point 😄. I think I actually would need one. I just ran it manually.
(Which definitely is bad form).

Thanks for the other comments too. It's nice to get some feedback 👍

Br.
Ville

ma 26. heinäk. 2021 klo 13.36 Martin Janiczek ***@***.***)
kirjoitti:

**Pull request closed without merging**